azuredevops/internal/service/wiki/resource_wiki_page.go

```diff
@@ -0,0 +1,169 @@
+*/
+var pageLock = sync.Mutex{}
+
+func ResourceWikiPage() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceWikiPageCreate,
+		Read:   resourceWikiPageRead,
+		Update: resourceWikiPageUpdate,
+		Delete: resourceWikiPageDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"project_id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.IsUUID,
+			},
+			"wiki_id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.IsUUID,
+			},
+			"path": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"content": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"etag": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceWikiPageCreate(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+
+	projectID := d.Get("project_id").(string)
+	wikiID := d.Get("wiki_id").(string)
+	path := d.Get("path").(string)
+	content := d.Get("content").(string)
+
+	pageLock.Lock()
+	defer pageLock.Unlock()
+
+	resp, err := clients.WikiClient.CreateOrUpdatePage(clients.Ctx, wiki.CreateOrUpdatePageArgs{
+		Parameters: &wiki.WikiPageCreateOrUpdateParameters{
+			Content: &content,
+		},
+		Project:        &projectID,
+		WikiIdentifier: &wikiID,
+		Path:           &path,
+	})
+
+	if err != nil {
+		return err
+	}
+	d.SetId(strconv.Itoa(*resp.Page.Id))
+	return resourceWikiPageRead(d, m)
+}
+
+func resourceWikiPageRead(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+	projectID := d.Get("project_id").(string)
+	wikiID := d.Get("wiki_id").(string)
+	path := d.Get("path").(string)
+	includeContent := true
+
+	resp, err := clients.WikiClient.GetPage(clients.Ctx, wiki.GetPageArgs{
+		Project:        &projectID,
+		WikiIdentifier: &wikiID,
+		Path:           &path,
+		IncludeContent: &includeContent,
+	})
+
+	if err != nil {
+		return err
+	}
+	etagValue := strings.Trim(strings.Join(*resp.ETag, " "), `\"`)
+	d.Set("etag", etagValue)
+	d.Set("content", *resp.Page.Content)
+	d.Set("path", *resp.Page.Path)
+	return nil
+}
+
+func resourceWikiPageUpdate(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+	projectID := d.Get("project_id").(string)
+	wikiID := d.Get("wiki_id").(string)
+
+	etag := d.Get("etag").(string)
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return err
+	}
+	content := d.Get("content").(string)
+
+	pageLock.Lock()
+	defer pageLock.Unlock()
+
+	_, err = clients.WikiClient.UpdatePageById(clients.Ctx, wiki.UpdatePageByIdArgs{
+		Parameters: &wiki.WikiPageCreateOrUpdateParameters{
+			Content: &content,
+		},
+		Project:        &projectID,
+		WikiIdentifier: &wikiID,
+		Id:             &id,
+		Version:        &etag,
+	})
+
+	if err != nil {
```

Comment:
Return the error directly, not need to handle the 404 error here. The should be handled in read.